### PR TITLE
Small tweaks, more tests

### PR DESCRIPTION
**Pull request by rocky**
"Catch[] needs to make sure it controls the expression evaluation
rather than having its expression get evaluated before it has a
change to wrap it in a `try`.

Internally, a plain `raise` is a bit more tidy Python wise
since the Python traceback will the path from raise to handler without
intervening tests of inappropriate exception handlers. This is a
matter of taste though.

Add more unit test tests.

### CODE REVIEW

**Comment by rocky** [comment]
It looks like there are still some bugs: 

```
Throw need not occur lexically inside Catch:

In[1]:=	f[x_] := (If[x < 0, Throw["negative"]]; Sqrt[x])	https://wolfram.com/xid/0y76co-bdikhw
In[2]:=	Catch[Sum[f[i], {i, 5, -5, -1}]]	https://wolfram.com/xid/0y76co-npb522
Out[2]=	negative (* comes out 0 instead *)
```

I'll look at this more later.

**Comment by rocky** [comment]
> It looks like there are still some bugs:
> 
> ```
> Throw need not occur lexically inside Catch:
> 
> In[1]:=	f[x_] := (If[x < 0, Throw["negative"]]; Sqrt[x])	
> In[2]:=	Catch[Sum[f[i], {i, 5, -5, -1}]]	
> Out[2]=	negative (* comes out 0 instead *)
> ```
> 
> I'll look at this more later.

File: mathics/builtin/control.py [hunk 1]
```diff
@@ -759,9 +760,8 @@ def apply1(self, expr, evaluation):
         except WLThrowInterrupt as e:
             return e.value
         return ret
- 
+
     def apply3(self, expr, form, f, evaluation):
-        'Catch[expr_, form_, f__:Identity]'
```
Comment (mmatera) [inline mathics/builtin/control.py:764]: Why did you remove this line? This was for stsblish the default handler function to the identity. Shouldn't it be there?

File: mathics/builtin/control.py [hunk 2]
```diff
@@ -771,7 +771,9 @@ def apply3(self, expr, form, f, evaluation):
             if match.is_true():
                 return Expression(f, e.value)
             else:
-                raise e
+                # A plain raise hide, this path and preserves the traceback
+                # of the call that was originally given.
+                raise
```
Comment (mmatera) [inline mathics/builtin/control.py:776]: OK, I es not aware of how this idiom works

**Review by mmatera — APPROVED** [review]

File: mathics/builtin/control.py (hunk 1 above)
Comment (rocky) [reply mathics/builtin/control.py:764]: Ooops - definitely a mistake.  It should be back in d0be8daf   Thanks for catching.

Comment (mmatera) [reply mathics/builtin/control.py:764]: OK. Then, let's merge this. Over the weekend I can try to figure out what happens with the throw inside called functions
